mypy/checkexpr.py

```diff
@@ -3108,112 +3108,135 @@ def infer_lambda_type_using_context(self, e: LambdaExpr) -> Tuple[Optional[Calla
-                    return
+                    return AnyType(TypeOfAny.from_error)
                 type_info = item.type
             elif isinstance(type_obj_type, AnyType):
-                return
+                return AnyType(TypeOfAny.from_another_any, source_any=type_obj_type)
             else:
                 self.msg.first_argument_for_super_must_be_type(type_obj_type, e)
-                return
+                return AnyType(TypeOfAny.from_error)
 
-            if isinstance(instance_type, (Instance, TupleType, TypeVarType)):
-                if isinstance(instance_type, TypeVarType):
-                    # Needed for generic self.
-                    instance_type = instance_type.upper_bound
-                    if not isinstance(instance_type, (Instance, TupleType)):
-                        # Too tricky, give up.
-                        # TODO: Do something more clever here.
-                        self.chk.fail(message_registry.UNSUPPORTED_ARG_2_FOR_SUPER, e)
-                        return
-                if isinstance(instance_type, TupleType):
-                    # Needed for named tuples and other Tuple[...] subclasses.
-                    instance_type = tuple_fallback(instance_type)
-                if type_info not in instance_type.type.mro:
-                    self.chk.fail(message_registry.SUPER_ARG_2_NOT_INSTANCE_OF_ARG_1, e)
-            elif isinstance(instance_type, TypeType) or (isinstance(instance_type, FunctionLike)
-                                                         and instance_type.is_type_obj()):
-                # TODO: Check whether this is a valid type object here.
-                pass
-            elif not isinstance(instance_type, AnyType):
+        if isinstance(instance_type, (Instance, TupleType, TypeVarType)):
+            simple_type = instance_type
+            if isinstance(simple_type, TypeVarType):
+                # Needed for generic self.
+                if isinstance(simple_type.upper_bound, (Instance, TupleType)):
+                    simple_type = simple_type.upper_bound
+                else:
+                    # Too tricky, give up.
+                    # TODO: Do something more clever here.
+                    self.chk.fail(message_registry.UNSUPPORTED_ARG_2_FOR_SUPER, e)
+                    return AnyType(TypeOfAny.from_error)
+            if isinstance(simple_type, TupleType):
+                # Needed for named tuples and other Tuple[...] subclasses.
+                simple_type = tuple_fallback(simple_type)
+            mro = simple_type.type.mro
+        elif isinstance(instance_type, FunctionLike) and instance_type.is_type_obj():
+            mro = instance_type.type_object().mro
+        elif isinstance(instance_type, TypeType):
+            item = instance_type.item
+            if isinstance(item, AnyType):
+                # Could be anything.
+                # XXX: TODO: Same as elif isinstance(instance_type, AnyType) below
```

Comment:
Oops, didn't do this yet

Reply:
No worries, just let us know when this ready for review.

Reply:
Should be ready now, thanks!